src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java

```diff
@@ -497,14 +497,47 @@ public LockItem acquireLock(final AcquireLockOptions options) throws LockNotGran
                     item.put(RECORD_VERSION_NUMBER, AttributeValue.builder().s(String.valueOf(recordVersionNumber)).build());
                     sortKeyName.ifPresent(sortKeyName -> item.put(sortKeyName, AttributeValue.builder().s(sortKey.get()).build()));
                     newLockData.ifPresent(byteBuffer -> item.put(DATA, AttributeValue.builder().b(SdkBytes.fromByteBuffer(byteBuffer)).build()));
+                    Optional<AtomicLong> lastTouchedAt = options.getClockSkewUpperBound()
+                      .map(unused -> new AtomicLong(System.currentTimeMillis()));
+                    lastTouchedAt.ifPresent(value -> item.put(LAST_TOUCHED_AT, AttributeValue.builder().s(String.valueOf(value.get())).build()));
+
+                    if (options.shouldSkipBlockingWait() && existingLock.isPresent() && !existingLock.get().isExpired()) {
+                        LockItem lockItem = existingLock.get();
+                        if (options.getClockSkewUpperBound().isPresent()) {
+                            // If we have provided an upper clock skew error bound then we rely on lastTouchedAt to know
+                            // whether the lock is expired without having to block.
+                            Long clockSkewUpperBound = options.getClockSkewUpperBound().get();
+                            if (lockItem.getLastTouchedAt().isPresent()) {
+                                long currentTime = System.currentTimeMillis();
+                                // If the lock hasn't been touched since the lease duration plus error bound then it's ours!
+                                if (currentTime > lockItem.getLastTouchedAt().get().get() + lockItem.getLeaseDuration() + clockSkewUpperBound) {
+                                    synchronized (lockItem) {
+                                        return upsertAndMonitorExpiredLock(options, key,
+                                          sortKey, deleteLockOnRelease, sessionMonitor,
+                                          existingLock, newLockData, item,
+                                          recordVersionNumber, lastTouchedAt);
+                                    }
+                                }
+                            }
+                        }
+
+                        /*
+                         * The lock is being held by someone and is still not expired. And the caller explicitly said not to perform a blocking wait;
+                         * We will throw back a lock not grant exception, so that the caller can retry if needed.
+                         *
+                         * Note that if we don't provide an upper clock skew error bound then we will NEED to wait to acquire a lock which is
+                         * expired but failed to be released/deleted due to ungraceful shutdown of the owning node.
+                         */
+                        throw new LockCurrentlyUnavailableException("The lock being requested is being held by another client.");
```

Comment:
To fix the bug I described above, you need to wrap this throw statement with a conditional:
```
if (!hasExpiredRecordRevisionNumber) {
    throw new LockCurrentlyUnavailableException("The lock being requested is being held by another client.");
}
```

And of course update the code comment.

Reply:
And add a unit test.

Reply:
Oh I see, that makes sense I think! 

Just as an fyi I've been busy with some medical stuff recently and haven't had time to address your comments but I really appreciate them. Will spend some time next week on this!

Reply:
Is

```java
if (!hasExpiredRecordRevisionNumber) {
    throw new LockCurrentlyUnavailableException("The lock being requested is being held by another client.");
}
```


equivalent to
```java
if (!lockTryingToBeAcquired.getRecordVersionNumber().equals(existingLock.get().getRecordVersionNumber())) {
    throw new LockCurrentlyUnavailableException("The lock being requested is being held by another client.");
}
```

? I wasn't sure what determines a record number being expired -- I think it would be when the local is different than the remote?

Being honest, I thought that
> This won't work and introduces a bug for backwards compatibility/those customers not using the new option. existingLock.get().isExpired() will always be false and this if statement will evaluate to true for people NOT using clockSkewUpperBound. Therefore, LockCurrentlyUnavailableException will be thrown whether the lock is expired or not.

was the case currently? e.g. if you don't blocking wait then you'll never be able to acquire a lock if it exists in dynamo

Reply:
> I'm working on getting approval from Amazon to publish open source code. Once I do that, I can add the unit test in another CR so that you can confidently fix your PR. The unit test has been missing for years.

I'll keep an eye out for this and rebase on top once merged!